src/log-pcap-stream.c

```diff
@@ -0,0 +1,312 @@
+#include "detect-engine-tag.h"
+#include "detect-tag-pcap.h"
+#include "log-pcap-stream.h"
+#include "output.h"
+#include "stream-tcp-reassemble.h"
+#include "threads.h"
+#include "threadvars.h"
+#include "tm-threads.h"
+
+#define MODULE_NAME "PcapLogStream"
+
+/**
+ *  \brief Variable to store the output context. Contains the filename of the
+ *   output file.
+ */
+typedef struct StreamOutputCtx_ {
+    char filename[NAME_MAX];
+    bool session_dump_enabled;
+} StreamOutputCtx;
+
+/**
+ *  \brief Variables to maintain context on this thread.
+ */
+typedef struct StreamPcapLogThreadData_ {
+    uint32_t count;
+    StreamOutputCtx *stream_output_ctx;
+} StreamPcapLogThreadData;
+
+/* Forward declarations for registration. */
+
+static int StreamPcapLog(ThreadVars *tv, void *thread_data, const Packet *p);
+static int StreamPcapLogCondition(ThreadVars *tv, const Packet *p);
+static TmEcode StreamPcapLogThreadInit(ThreadVars *tv, const void *initdata,
+        void **data);
+static TmEcode StreamPcapLogThreadDeInit(ThreadVars *tv, void *thread_data);
+static OutputInitResult StreamPcapLogInitCtx(ConfNode *conf);
+static void StreamPcapLogFileDeInitCtx(OutputCtx *output_ctx);
+
+static StreamOutputCtx *GetStreamOutputCtx(const OutputCtx *output_ctx);
+static void GeneratePcapFiles(ThreadVars *tv, StreamPcapLogThreadData *td,
+        const Packet *p);
+
+/**
+ * \brief Stream pcap logging main function.
+ * \param tv thread-specific variables.
+ * \param thread_data thread module specific data.
+ * \param p Pointer to current packet being processed.
+ * \return TM_ECODE_OK on success.
+ */
+static int StreamPcapLog(ThreadVars *tv, void *thread_data, const Packet *p)
+{
+    StreamPcapLogThreadData *td = (StreamPcapLogThreadData *) thread_data;
+    td->count++;
+    GeneratePcapFiles(tv, thread_data, p);
+    return TM_ECODE_OK;
+}
+
+/**
+ * \brief Generate pcap events for the alerted packet. Handles single packet
+ *  protocols/alerts and multiple packet alerts. For single packet
+ *  alerts, the alert file or event message are created and output.
+ *  For multiple packet alerts, the pcap file is created.
+ * \param tv thread-specific variables.
+ * \param td thread data containing the output context.
+ * \param p Pointer to current packet being processed.
+ */
+static void GeneratePcapFiles(ThreadVars *tv, StreamPcapLogThreadData *td,
+        const Packet *p)
+{
+    if (!(p->flags & PKT_HAS_FLOW) && p->flags & PKT_HAS_TAG) {
+        /*
+         * Single packet protocols (eg. ICMP) won't have a flow created for
+         * them. A tagged rule alert can still trigger on the packet meaning
+         * we'd want to produce a PCAP.
+         */
+        for (uint16_t x = 0; x < p->alerts.cnt; x++) {
+            /* Is the alert from a tagged signature? */
+            if (p->alerts.alerts[x].s->sm_arrays[DETECT_SM_LIST_TMATCH] !=
+                NULL) {
+                TaggedPcapEntry *pcap_file = SetupTaggedPcap(p, p->alerts
```

Comment:
it looks like `SetupTaggedPcap` can return NULL in error cases, which would crash on the next line.

Reply:
Good catch thanks. Not sure how I missed actually handling the returned NULL after working out the error cases... I will add a check error and log and return in the case of an error.